model/riscv_crypto_entropy_source.sail

```diff
@@ -0,0 +1,90 @@
+
+/*
+ * file: riscv_crypto_entropy_source.sail
+ *
+ * Contains a basic model for the RISC-V cryptography extension entropy
+ * source. See section 4 of the specifications for details.
+ * - https://github.com/riscv/riscv-crypto/releases
+ *
+ */
+
+/*
+ * The mnoise CSR
+ * ------------------------------------------------------------
+ */
+
+register mnoise : xlenbits
+
+/* mnoise is defined in machine mode only */
+function clause ext_is_CSR_defined (0x7A9, Machine) = {
+    true
+}
+
+function clause ext_read_CSR(0x7A9) = {
+    Some(mnoise)
+}
+
+/* Only bit 31 of mnoise is architecturally defined. */
+function clause ext_write_CSR(0x7A9, value) = {
+    mnoise[31] = value[31];
+    Some(mnoise)
+}
+
+/* Used by mentropy to check if we are (not) in noise capture mode */
+val      in_noise_capture_mode : unit -> bool effect {rreg}
+function in_noise_capture_mode() = {
+    if(bit_to_bool(mnoise[31])) then true else false
+}
+
+/*
+ * The mentropy CSR
+ * ------------------------------------------------------------
+ */
+
+/* Valid return states for reading the mentropy CSR.
+enum mentropy_opst = {
+    BIST, // Built-in-self-test. No randomness sampled.
+    ES16, // Entropy-sample-16. Valid 16-bits of randomness sampled.
+    WAIT, // Device still gathering entropy.
+    DEAD  // Fatal device compromise. No randomness sampled.
+}
+
+/* Mapping of status codes and their actual encodings. */
+mapping opst_code : mentropy_opst <-> bits(2) = {
+    BIST <-> 0b00,
+    ES16 <-> 0b01,
+    WAIT <-> 0b10,
+    DEAD <-> 0b11
+}
+
+/* 
+ * Dummy function representing an entropy samping.
+ * TODO: Examine how best to support this in SAIL?
```

Comment:
You'll likely need a Platform.get_random_bits for OCaml and a wrapper around getrandom for C that this then exposes.

Reply:
Is this process documented somewhere? Or is there an example in another architecture I can use as a reference? It might be a bit beyond my OCaml abilities at the moment.

For context - there is a lot of worry/discussion about how a mechanism like this gets tested, particularly for the architectural compliance tests. E.g., changing exactly what the tests check, or switching in a deterministic PRNG for testing etc.

Would it be okay to leave this as a TODO for the purposes of getting the rest of this PR merged (once I've fixed the various other issues) and coming back to it? I'll understand if not.

Reply:
Search for `ocaml:` in model/ to find all the existing "FFI bindings" and pick one that looks close to what you need this to do. I can see the value in determinism, but it probably should be configurable.

Reply:
Thanks for the pointer, I'll give it a whirl.

I agree it should be configurable. I expect it will end up being modified as the standardisation / compliance processes are discussed more. The aim at the moment I guess is just to get something minimally functional in there.